c2rust-analyze/src/rewrite/expr/distribute.rs

```diff
@@ -92,19 +105,27 @@ pub fn distribute<'tcx>(
                     loc,
                     sub_loc: key.1,
                     desc: origin.desc,
+                    priority,
                 });
         }
     }
 
-    // If a single `HirId` has rewrites from multiple different pieces of MIR, it's ambiguous how
-    // to order those rewrites.  (`mir_rewrites` only establishes an ordering between rewrites on
-    // the same `Location`.)  For now, we complain if we see this ambiguity; in the future, we may
-    // need to add rules to resolve it in a particular way, such as prioritizing one `SubLoc` or
-    // `MirOriginDesc` over another.
-    for (&hir_id, infos) in &info_map {
-        let all_same_loc = infos.iter().skip(1).all(|i| {
-            i.loc == infos[0].loc && i.sub_loc == infos[0].sub_loc && i.desc == infos[0].desc
-        });
+    // If a single `HirId` has rewrites from multiple different pieces of MIR at the same
+    // `Priority`, it's ambiguous how to order those rewrites.  (`mir_rewrites` only establishes an
+    // ordering between rewrites on the same `Location`.)  For now, we complain if we see this
+    // ambiguity; in the future, we may need to add rules to resolve it in a particular way, such
+    // as prioritizing one `SubLoc` over another.
+    for (&hir_id, infos) in &mut info_map {
```

Comment:
How about this?

```rust
    for (&hir_id, infos) in &info_map {
        for (priority, infos) in infos.iter().into_group_map_by(|i| i.priority) {
            if !infos
                .iter()
                .map(|i| (i.loc, &i.sub_loc, i.desc))
                .all_equal()
            {
                info!("rewrite info of {priority:?}:");
                for i in infos {
                    info!("  {:?}, {:?}, {:?}: {:?}", i.loc, i.sub_loc, i.desc, i.rw);
                }
                let ex = tcx.hir().expect_expr(hir_id);
                error!(
                    "multiple distinct locations produced rewrites for {:?} {:?}",
                    ex.span, ex,
                );
            }
        }
    }
```

Using `.into_group_map_by` and `.all_equal` would make things a lot more explicit/declarative, and this way we also move the `error!` to the priority group that is ambiguous, instead of printing out all the other infos, too.